src/lib/datepicker/calendar.ts

```diff
@@ -97,57 +107,181 @@ export class MdCalendar implements AfterContentInit {
+        break;
+      case HOME:
+        this._activeDate = this._monthView ?
+            new SimpleDate(this._activeDate.year, this._activeDate.month, 1) :
+            this._addCalendarMonths(this._activeDate, -this._activeDate.month);
+        break;
+      case END:
+        this._activeDate = this._monthView ?
+            new SimpleDate(this._activeDate.year, this._activeDate.month + 1, 0) :
+            this._addCalendarMonths(this._activeDate, 11 - this._activeDate.month);
+        break;
+      case PAGE_UP:
+        if (event.altKey) {
+          this._activeDate = this._addCalendarYears(this._activeDate, this._monthView ? -1 : -10);
+        } else {
+          this._activeDate = this._monthView ?
+              this._addCalendarMonths(this._activeDate, -1) :
+              this._addCalendarYears(this._activeDate, -1);
+        }
+        break;
+      case PAGE_DOWN:
+        if (event.altKey) {
+          this._activeDate = this._addCalendarYears(this._activeDate, this._monthView ? 1 : 10);
+        } else {
+          this._activeDate = this._monthView ?
+              this._addCalendarMonths(this._activeDate, 1) :
+              this._addCalendarYears(this._activeDate, 1);
+        }
+        break;
+      case ENTER:
+        if (this._monthView) {
+          if (this._dateFilterForViews(this._activeDate)) {
+            this._dateSelected(this._activeDate);
+          }
+        } else {
+          this._monthSelected(this._activeDate);
+        }
+        break;
+      default:
+        // Don't prevent default on keys that we don't explicitly handle.
+        return;
+    }
+
+    event.preventDefault();
+  }
+
+  /** Adds the given number of days to the date. */
+  private _addCalendarDays(date: SimpleDate, days: number): SimpleDate {
+    return date.add({'days': days});
+  }
+
+  /**
+   * Adds the given number of months to the date. Months are counted as if flipping pages on a
+   * calendar and then finding the closest date in the new month. For example when adding 1 month to
+   * Jan 31 2017 the resulting date will be Feb 28 2017.
+   */
+  private _addCalendarMonths(date: SimpleDate, months: number): SimpleDate {
+    let newDate = date.add({'months': months});
+
+    // It's possible to wind up in the wrong month if the original month has more days than the new
+    // month. In this case we want to go to the last day of the desired month.
+    // Note: the additional + 12 % 12 ensures we end up with a positive number, since JS % doesn't
+    // guarantee this.
+    if (newDate.month != ((date.month + months) % 12 + 12) % 12) {
+      newDate = new SimpleDate(newDate.year, newDate.month, 0);
+    }
+
+    return newDate;
+  }
+
+  /**
+   * Adds the given number of months to the date. Months are counted as if flipping 12 pages for
+   * each year on a calendar and then finding the closest date in the new month. For example when
+   * adding 1 year to Feb 29 2016 the resulting date will be Feb 28 2017.
```

Comment:
Ultra nit: add comma after 'Feb 29, 2016' (same for line 241)? Love the detailed comment though